keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2150 @@
+  DeletionTimestamp. Otherwise, allocation gets aborted immediately.
+
+  What this avoids is the situation where an allocation succeed without having
+  an object where the result can be stored. The driver can also be killed at
+  any time: when it restarts, the finalizer indicates that allocation may be in
+  progress and has to be completed or aborted.
+
+- In a cluster with multiple scheduler instances, two pods might get
+  scheduled concurrently by different schedulers. When they reference
+  the same ResourceClaim which may only get used by one pod at a time,
+  only one pod can be scheduled.
+
+  Both schedulers try to add their pod to the `ReservedFor` field, but only the
+  update that reaches the API server first gets stored. The other one fails
+  with a conflict error and the scheduler which issued it knows that it must
+  put the pod back into the queue, waiting for the ResourceClaim to become
+  usable again.
+
+### Custom parameters
+
+To support arbitrarily complex parameters, both ResourceClass and ResourceClaim
+contain one field of type
+[v1.TypedLocalObjectReference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#typedlocalobjectreference-v1-core)
+which references a separate object. This information is sufficient for generic clients to
+retrieve the parameters. For ResourceClass, that object must be
+cluster-scoped. For ResourceClaim, it must be in the same namespace as the
+ResourceClaim and thus the Pod. Which objects a resource driver accepts as parameters depends on
+the driver.
+
+This approach was chosen because then validation of the parameters can be done
+with a CRD and that validation will work regardless of where the parameters
+are needed.
+
+A resource driver may support modification of the parameters while a resource
+is in use ("online resizing"). It may update the ResourceClaim status to
+reflect the modified state, for example by increasing the number of concurrent
+users. However, the state must not be modified such that a user of the resource
+no longer has access.
+
+Parameters may get deleted before the ResourceClaim or ResourceClass that
+references them. In that case, a pending resource cannot be allocated until the
+parameters get recreated. An allocated resource must remain usable and freeing
+it must be possible. To support this, resource drivers must copy all relevant
+information into the ResourceClaim status when allocating it.
+
+### Allocation modes
+
+Allocation of a resource happens either immediately when a ResourceClaim gets
+created (“immediate allocation”) or when a Pod is getting scheduled which
+needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec.
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
```

Comment:
should we start with just a single Pod and iterate on support for multiple pods later?

Reply:
I suspect that if we do that, then we will design something that cannot be extended later to support multiple pods. We need to keep that requirement in mind already now when designing the solution.